pkg/agent/nodeportlocal/k8s/k8s_controller.go

```diff
@@ -0,0 +1,165 @@
+// +build !windows
+
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8s
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/vmware-tanzu/antrea/pkg/agent/nodeportlocal/portcache"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+)
+
+const (
+	minRetryDelay = 2 * time.Second
+	maxRetryDelay = 120 * time.Second
+	numWorkers    = 4
+)
+
+type Controller struct {
+	portTable  *portcache.PortTable
+	kubeClient clientset.Interface
+	queue      workqueue.RateLimitingInterface
+	Ctrl       cache.Controller
+	CacheStore cache.Store
+	PodToIP    map[string]string
+}
+
+func NewNPLController(kubeClient clientset.Interface, pt *portcache.PortTable) *Controller {
+	nplCtrl := Controller{
+		kubeClient: kubeClient,
+		portTable:  pt,
+		PodToIP:    make(map[string]string),
+	}
+	nplCtrl.queue = workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "nodeportlocal")
+	return &nplCtrl
+}
+
+// Run starts to watching and process Pod updates for the Node where Antrea Agent is running.
+// It starts a fixed numbers of queues and one worker per queue to process the objects.
+// Each object is mapped to one queue based on the hash of the object key. This ensures that
+// update for one object always gets processes by the same worker.
```

Comment:
comment is no longer accurate. There is a single workqueue but the workqueue implementation does not allow multiple workers to process the same item simultaneously.

> a single item will not be processed multiple times concurrently
https://godoc.org/k8s.io/client-go/util/workqueue